File: AKS.py
import math
import array as arr
from fractions import gcd

def perfectPower(n):
    """Checks if number is a power of another integer, 
       if it returns true, then it is composite.
    """
    for b in range(2,int(math.log2(n))+1):       
        a=n**(1/b)
        if a-int(a) == 0:    
            return(True)    
    return(False)

def findR(n):
   """Find smallest r such that the order of n mod r > log2(n)^2.
   """
   maxK = math.log2(n)**2   
   maxR = math.log2(n)**5   
   nexR = True              
   r = 1                   
   while nexR == True:
       r +=1
       nexR = False
       k = 0
       while k <= maxK and nexR == False:
           k = k+1
           if fastMod(n,k,r) == 0 or fastMod(n,k,r) == 1:
               nexR = True
   return(r)

def fastMod(base,power,n):
    """Implement fast modular exponentiation.
    """
    r=1
    while power > 0:
        if power % 2 == 1:
            r = r * base % n
        base = base**2 % n
        power = power // 2
    return(r)

def fastPoly(base,power,r):
    """Use fast modular exponentiation for polynomials to raise them to a big power.
    """
    x = arr.array('d',[],)
    for i in range(len(base)):
        x.append(0)
    x[(0)] = 1 
    n = power
    
    while power > 0:
        if power % 2 == 1: 
            x = multi(x,base,n,r)
        base = multi(base,base,n,r)
        power = power // 2        
    return(x)

def multi(a,b,n,r):
    """Function used by fastPoly to multiply two polynomials together.
    """ 
    x = arr.array('d',[])
    for i in range(len(a)+len(b)-1):
        x.append(0)
    for i in range(len(a)):
        for j in range(len(b)):
            x[(i+j) % r ] += a[(i)] * b[(j)] 
            x[(i+j) % r] = x[(i+j) % r] % n 
    for i in range(r,len(x)):
            x=x[:-1]
    return(x)

def eulerPhi(r):
    """Implement the euler phi function
    """
    x = 0        
    for i in range(1, r + 1):
        if math.gcd(r, i) == 1:
            x += 1
    return x


def aks(n):
    """ The main AKS algorithm
    """
    if perfectPower(n) == True:                     #step 1
        return('Composite')
    
    r = findR(n)                                    #step 2

    for a in range(2,min(r,n)):                     #step 3
        if math.gcd(a,n) > 1:                       
            return('Composite')

    if n <= r:                                      #step 4
        return('Prime')

    x = arr.array('l',[],)                         #step 5
    for a in range(1,math.floor((eulerPhi(r))**(1/2)*math.log2(n))):      
        x = fastPoly(arr.array('l',[a,1]),n,r)
        x[(0)] = x[(0)] - a % n
        x[(n % r )] = x[(n % r )] - 1
        if  any(x):
            return('Composite')
    return('Prime')                                 #step 6

File: AKS.go
package main

import (
	"fmt"
	"math"
	"math/big"
)

func main() {
	var n int64 = 1009

	fmt.Println(aks(int64(n)))
}

func aks(n int64) string {
	if perfectPower(n) == true { 					 //step 1
		return ("Composite")
	}

	r := findR(n)                                                    //step 2
	for a := 2; float64(a) < math.Min(float64(r), float64(n)); a++ { //step 3
		if GCD(a, int(n)) > 1 {
			return "Composite"
		}
	}
	if n <= r { //step 4
		return "Prime"
	}

	var x = []*big.Int{} 						 //step 5
	var a float64 = 1
	for a < math.Floor(math.Pow(eulerPhi(r), 1/2)*math.Log2(float64(n))) {
		x = fastPoly([]*big.Int{big.NewInt(1), big.NewInt(1)}, n, r)
		if notZero(x, n) {
			return "Coprime"
		}
		a++
	}
	return "Prime"							 //step 6

}

func notZero(a []*big.Int, n int64) bool {
	for i := 0; i < len(a); i++ {
		if (a[i].Mod(a[i], big.NewInt(n))).Int64() != 0 {
			return true
		}
	}
	return false
}

func perfectPower(n int64) bool {
	var i float64 = 2
	for i < math.Log2(float64(n))+1 {
		var a float64
		a = math.Pow(float64(n), float64(1/i))
		if a == float64(int64(a)) {
			return true
		}
		i++
	}
	return false
}

func findR(n int64) int64 {
	maxK := math.Pow(math.Log2(float64(n)), 2)
	nextR := true
	var r int64 = 1

	for nextR == true {
		r++
		nextR = false
		var k int = 0
		for float64(k) <= maxK && nextR == false {
			k++
			if fastMod(n, k, r) == 0 || fastMod(n, k, r) == 1 {
				nextR = true
			}
		}
	}
	return r
}

func fastMod(base int64, power int, n int64) int {
	var r int = 1
	for power > 0 {
		if int(power)%2 == 1 {
			r = int(r) * int(base) % int(n)
		}
		base = int64(int(math.Pow(float64(base), 2)) % int(n))
		power = power / 2
	}
	return r
}

func fastPoly(base []*big.Int, power int64, r int64) []*big.Int {
	var x = make([]*big.Int, int(r))
	for i := range x {
		x[i] = big.NewInt(0)
	}

	a := base[0]

	x[0] = big.NewInt(1)
	n := power

	for power > 0 {
		if int(power)%2 == 1 {
			x = multi(x, base, n, r)
		}
		base = multi(base, base, n, r)
		power = power / 2
	}

	x[0].Sub(x[0], a)
	x[n%int64(r)].Sub(x[n%int64(r)], big.NewInt(1))

	return x
}

func multi(a []*big.Int, b []*big.Int, n int64, r int64) []*big.Int {
	var temp = big.NewInt(0)
	var x = make([]*big.Int, int(r))
	for i := range x {
		x[i] = big.NewInt(0)
	}
	for i := 0; i < len(a); i++ {
		for j := 0; j < len(b); j++ {
			if (i + j) < int(r) {
				temp.Mul(a[i], b[j])
				x[i+j].Add(x[i+j], temp)
			} else {
				temp.Mul(a[i], b[j])
				x[(i+j)%(int(r))].Add(x[(i+j)%(int(r))], temp)
			}
		}
	}
	return x
}

func eulerPhi(r int64) float64 {
	var x float64 = 0
	for i := 0; i < int(r)+1; i++ {
		if GCD(int(r), i) == 1 {
			x++
		}
	}
	return x
}

// GCD : greatest common divisor (GCD) via Euclidean algorithm
func GCD(a, b int) int {
	for b != 0 {
		t := b
		b = a % b
		a = t
	}
	return a
}
